### ISSUE TITLE
[Bug] Qwen3 SFT loss-mask generator produces token IDs inconsistent with chat template for consecutive tool responses

### ISSUE THREAD

**Issue by HelanHu** [THUDM/slime#2263]
### Bug Description

slime 的 Qwen3 多轮 SFT 会错误编码并行工具调用的返回结果，导致训练时使用的 token 序列与 Qwen chat template 在推理时生成的 token 序列不一致。

问题发生在以下合法且常见的并行工具调用流程中：

```text
assistant 同时发起 tool call A 和 B
→ tool A 返回 result A
→ tool B 返回 result B
→ assistant 读取两个结果并继续
```

在 OpenAI 风格的结构化消息中，每个工具结果分别保存为一条 `tool` 消息，因此 result A 和 result B 是两条连续的 `tool` 消息。它们属于同一批并行调用的返回结果，不是两个独立的 user turn。

Qwen chat template 能识别这种结构：它会把连续的 `tool` 消息放进同一个 user turn，并在其中生成多个 `<tool_response>`。但是，slime 的 `gen_multi_turn_loss_mask_qwen3()` 为了构造 loss mask，会把每条消息单独交给 chat template 渲染。这造成每个工具结果都包装成独立的 user turn，连续的工具结果会产生连续的user turn。

```text
<|im_start|>user
<tool_response>
contents of a.py
</tool_response><|im_end|>
<|im_start|>user
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

slime 在两个工具结果之间错误插入了一组 `<|im_end|><|im_start|>user` token。**更重要的是，这个函数返回的 token IDs 会被直接用作 SFT 的 `sample.tokens`，模型实际训练的输入序列也被改变了。**

推理时，如果使用同一个 Qwen tokenizer 对完整消息历史正常调用 `apply_chat_template()`，两个工具结果会被正确合并。于是模型训练时看到的是“两个 user turn”，推理时看到的是“一个 user turn 中的两个 tool responses”，构成训练/推理 tokenization 不一致。

**本质上，qwen3 loss-mask generator 消息逐条渲染的设计没有注意处理连续 tool messages 对相邻消息上下文的依赖，破坏了模板原本对连续 tool messages 的上下文处理逻辑。训练序列因此多出了推理时不存在的 turn boundary，最终造成训推不一致。**

### Steps to Reproduce

以下复现只需要 tokenizer，不需要 GPU 或模型权重。

```python
from transformers import AutoTokenizer

from slime.utils.mask_utils import MultiTurnLossMaskGenerator


model_name = "Qwen/Qwen3-30B-A3B-Instruct-2507"
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)

tools = [
    {
        "type": "function",
        "function": {
            "name": "read_file",
            "description": "Read a file.",
            "parameters": {
                "type": "object",
                "properties": {"path": {"type": "string"}},
                "required": ["path"],
            },
        },
    }
]

messages = [
    {"role": "system", "content": "You are a coding assistant."},
    {"role": "user", "content": "Read a.py and b.py."},
    {
        "role": "assistant",
        "content": "I will read both files.",
        "tool_calls": [
            {
                "id": "call_a",
                "type": "function",
                "function": {
                    "name": "read_file",
                    "arguments": '{"path":"a.py"}',
                },
            },
            {
                "id": "call_b",
                "type": "function",
                "function": {
                    "name": "read_file",
                    "arguments": '{"path":"b.py"}',
                },
            },
        ],
    },
    {
        "role": "tool",
        "name": "read_file",
        "tool_call_id": "call_a",
        "content": "contents of a.py",
    },
    {
        "role": "tool",
        "name": "read_file",
        "tool_call_id": "call_b",
        "content": "contents of b.py",
    },
    {"role": "assistant", "content": "Done."},
]

expected_token_ids = tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)

generator = MultiTurnLossMaskGenerator(tokenizer, tokenizer_type="qwen3")
actual_token_ids, loss_mask = generator.get_loss_mask(messages, tools=tools)

print("expected length:", len(expected_token_ids))
print("actual length:  ", len(actual_token_ids))
print("token IDs match:", actual_token_ids == expected_token_ids)

assert len(actual_token_ids) == len(loss_mask)
assert actual_token_ids == expected_token_ids
```

最后一个断言在当前 `main` 上失败。关键差异如下：

```text
# Expected: 两个结果位于同一个 user turn
<|im_start|>user
<tool_response>
contents of a.py
</tool_response>
<tool_response>
contents of b.py
</tool_response><|im_end|>

# Actual: 中间额外插入了 turn boundary
<|im_start|>user
<tool_response>
contents of a.py
</tool_response><|im_end|>
<|im_start|>user
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

该问题的最小触发条件是消息历史中存在两条连续的 `role="tool"` 消息。单个工具调用不会触发；两个工具如果是串行调用，消息顺序为 `assistant → tool → assistant → tool`，也不会触发，因为两个工具结果本来就属于不同的对话轮次。

### Expected Behavior

对于合法的 Qwen 对话，SFT loss-mask generator 返回的 token IDs 应与目标 tokenizer 对完整对话的渲染结果一致：

```python
actual_token_ids == tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)
```

loss mask 可以在该 token 序列上单独构造，但构造 loss mask 不应改变模型输入序列。

对于复现代码中的两个并行工具结果，预期渲染结果是一个 user turn 中包含两个 `<tool_response>`：

```text
<|im_start|>user
<tool_response>
contents of a.py
</tool_response>
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

也就是说，结构化数据中虽然有两条 `tool` 消息，但在 Qwen 模型输入中，它们应属于同一个工具结果轮次。

### Actual Behavior

`gen_multi_turn_loss_mask_qwen3()` 没有一次性渲染完整消息历史，而是循环处理每条消息：

```python
for message in messages:
    tokenizer.apply_chat_template([prefix_message, message], ...)
```

对于第一条 tool result，函数实际调用的模板输入中只有这一条 tool 消息；对于第二条也是如此。模板无法判断两条消息相邻，因此两次调用都会各自生成完整的 user turn。

实际训练序列变成：

```text
<|im_start|>user
<tool_response>
contents of a.py
</tool_response><|im_end|>
<|im_start|>user
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

相较于预期结果，中间多出了：

```text
<|im_end|>
<|im_start|>user
```

`sft_rollout.generate_rollout()` 随后执行：

```python
token_ids, loss_mask = MASK_GENERATOR.get_loss_mask(messages, tools=tools)
sample.tokens = token_ids
```

因此这些多余 token 会真正进入训练，而不只是影响 loss mask 的标记位置。推理时完整渲染同一段消息历史不会产生这些 token，所以模型消费并行工具结果时的训练上下文与推理上下文不同。

作为额外验证，在不更换 tokenizer 和输入数据的情况下，仅将 slime 的 loss-mask 路径从 `qwen3` 换成 `qwen3_5`，复现代码得到的 token IDs 就与完整 `apply_chat_template()` 结果一致。`qwen3_5` 路径会先渲染完整对话，再根据完整文本定位 assistant token。这个对照表明，问题来自 `qwen3` 路径的逐消息渲染方式，而不是 Qwen chat template 不支持并行工具结果，也不是结构化消息格式错误。

### Environment

- slime version: `06ffdbe22be068b52f9ed0fc318c473f7030197e` (`main`)
- Python version: `3.13.13`
- Transformers version: `4.57.6`
- Model tokenizer: `Qwen/Qwen3-30B-A3B-Instruct-2507`
- PyTorch version: N/A（仅使用 tokenizer 复现）
- CUDA/ROCm version: N/A
- GPU type and count: N/A
- OS: Linux
- SGLang version: N/A
- Megatron-LM version: N/A

### Additional Context

仓库现有的 `test_loss_mask_qwen3_tools` 已经构造了相同的消息结构：

```text
assistant(tool_call A, tool_call B)
→ tool(result A)
→ tool(result B)
→ assistant
```

但是该测试目前只验证：

```python
len(all_token_ids) == len(all_loss_masks)
```

这只能证明 token 数量与 mask 数量相同，不能证明 token 序列本身正确。当前实现同时生成了错误 token 和等长的 mask，因此现有断言仍然通过。

如果增加以下 oracle 对比，就可以捕获该问题：

```python
expected_token_ids = tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)
assert all_token_ids == expected_token_ids
```

我们还从真实多轮工具数据中取了 200 条轨迹。对每条轨迹，分别生成两组 token IDs：一组由目标 tokenizer 对完整 `messages + tools` 调用 `apply_chat_template()` 得到，作为 Qwen chat template 的预期结果；另一组由 slime 的 `qwen3` loss-mask generator 得到，也就是实际进入 SFT 的 token IDs。逐 token 比较后，123 条完全一致，77 条不一致。出现不一致的 77 条轨迹全部包含连续的 tool messages，而另外 123 条均不包含；这与最小复现得到的触发条件完全一致。

---

### Bug Description

The Qwen3 multi-turn SFT path tokenizes results from parallel tool calls incorrectly. Consequently, the token sequence used for training differs from the sequence produced by the Qwen chat template at inference time.

The bug occurs in a valid and common parallel tool-call flow:

```text
assistant issues tool calls A and B together
→ tool A returns result A
→ tool B returns result B
→ assistant consumes both results and continues
```

In an OpenAI-style message history, each result is represented by its own `tool` message. The two consecutive tool messages belong to one batch of parallel results, not to two separate user turns.

The Qwen chat template handles this by placing consecutive tool messages in one user turn containing multiple `<tool_response>` blocks. However, `gen_multi_turn_loss_mask_qwen3()` renders each message separately while constructing the loss mask. Each isolated tool message is therefore wrapped in its own user turn, inserting an extra `<|im_end|><|im_start|>user` boundary between the two results.

The returned token IDs are assigned directly to `sample.tokens`, so this changes the actual SFT input rather than only the mask. When the same full message history is rendered normally at inference time, the tool results are grouped correctly. The model is therefore trained and inferred with different tokenizations of the same history. In short, the current `qwen3` path produces a training sequence that diverges from the Qwen chat template's expected rendering of the same structured messages, causing the train/inference mismatch.

### Steps to Reproduce

The reproduction only requires the tokenizer; no GPU or model weights are needed.

```python
from transformers import AutoTokenizer

from slime.utils.mask_utils import MultiTurnLossMaskGenerator


model_name = "Qwen/Qwen3-30B-A3B-Instruct-2507"
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)

tools = [
    {
        "type": "function",
        "function": {
            "name": "read_file",
            "description": "Read a file.",
            "parameters": {
                "type": "object",
                "properties": {"path": {"type": "string"}},
                "required": ["path"],
            },
        },
    }
]

messages = [
    {"role": "system", "content": "You are a coding assistant."},
    {"role": "user", "content": "Read a.py and b.py."},
    {
        "role": "assistant",
        "content": "I will read both files.",
        "tool_calls": [
            {
                "id": "call_a",
                "type": "function",
                "function": {
                    "name": "read_file",
                    "arguments": '{"path":"a.py"}',
                },
            },
            {
                "id": "call_b",
                "type": "function",
                "function": {
                    "name": "read_file",
                    "arguments": '{"path":"b.py"}',
                },
            },
        ],
    },
    {
        "role": "tool",
        "name": "read_file",
        "tool_call_id": "call_a",
        "content": "contents of a.py",
    },
    {
        "role": "tool",
        "name": "read_file",
        "tool_call_id": "call_b",
        "content": "contents of b.py",
    },
    {"role": "assistant", "content": "Done."},
]

expected_token_ids = tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)

generator = MultiTurnLossMaskGenerator(tokenizer, tokenizer_type="qwen3")
actual_token_ids, loss_mask = generator.get_loss_mask(messages, tools=tools)

print("expected length:", len(expected_token_ids))
print("actual length:  ", len(actual_token_ids))
print("token IDs match:", actual_token_ids == expected_token_ids)

assert len(actual_token_ids) == len(loss_mask)
assert actual_token_ids == expected_token_ids
```

The final assertion fails on current `main`. The relevant difference is:

```text
# Expected: both results are in one user turn
<|im_start|>user
<tool_response>
contents of a.py
</tool_response>
<tool_response>
contents of b.py
</tool_response><|im_end|>

# Actual: an extra turn boundary is inserted
<|im_start|>user
<tool_response>
contents of a.py
</tool_response><|im_end|>
<|im_start|>user
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

The minimal trigger is two consecutive messages with `role="tool"`. A single tool result does not trigger the bug. Two serial calls also do not trigger it because their message order is `assistant → tool → assistant → tool`; those results correctly belong to separate turns.

### Expected Behavior

For a valid Qwen conversation, the token IDs returned by the SFT loss-mask generator should match the target tokenizer's full-conversation rendering:

```python
actual_token_ids == tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)
```

The loss mask may be constructed over this sequence, but constructing it must not change the model input tokens.

For the two parallel results in the reproduction, the expected rendering is one user turn containing two `<tool_response>` blocks:

```text
<|im_start|>user
<tool_response>
contents of a.py
</tool_response>
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

### Actual Behavior

`gen_multi_turn_loss_mask_qwen3()` renders messages one at a time instead of rendering the full history:

```python
for message in messages:
    tokenizer.apply_chat_template([prefix_message, message], ...)
```

Neither isolated tool message can observe the adjacent tool message, so both template calls produce a complete user turn. The training sequence becomes:

```text
<|im_start|>user
<tool_response>
contents of a.py
</tool_response><|im_end|>
<|im_start|>user
<tool_response>
contents of b.py
</tool_response><|im_end|>
```

Compared with the expected output, it contains the additional boundary:

```text
<|im_end|>
<|im_start|>user
```

`sft_rollout.generate_rollout()` then assigns these IDs directly to the training sample:

```python
token_ids, loss_mask = MASK_GENERATOR.get_loss_mask(messages, tools=tools)
sample.tokens = token_ids
```

The extra boundary tokens therefore enter training. Normal full-history rendering at inference does not produce them, so the model receives different contexts when consuming parallel tool results during training and inference.

As a control, keeping the same Qwen3 tokenizer and input data but using slime's `qwen3_5` full-conversation/offset-based path produces token IDs identical to `apply_chat_template()`. This indicates that the issue is caused by per-message rendering in the `qwen3` path, not by the Qwen template or the structured tool-call format.

### Environment

- slime version: `06ffdbe22be068b52f9ed0fc318c473f7030197e` (`main`)
- Python version: `3.13.13`
- Transformers version: `4.57.6`
- Model tokenizer: `Qwen/Qwen3-30B-A3B-Instruct-2507`
- PyTorch version: N/A (tokenizer-only reproduction)
- CUDA/ROCm version: N/A
- GPU type and count: N/A
- OS: Linux
- SGLang version: N/A
- Megatron-LM version: N/A

### Additional Context

The existing `test_loss_mask_qwen3_tools` already constructs the same message shape:

```text
assistant(tool_call A, tool_call B)
→ tool(result A)
→ tool(result B)
→ assistant
```

However, it currently checks only:

```python
len(all_token_ids) == len(all_loss_masks)
```

The implementation produces an incorrect token sequence and a mask of the same length, so this assertion still passes. Comparing the generated IDs with the tokenizer's full-conversation output would detect the bug:

```python
expected_token_ids = tokenizer.apply_chat_template(
    messages,
    tools=tools,
    tokenize=True,
    add_generation_prompt=False,
)
assert all_token_ids == expected_token_ids
```

We also sampled 200 real multi-turn tool trajectories. For each trajectory, we compared two token-ID sequences: (1) the expected sequence produced by applying the target tokenizer's `apply_chat_template()` once to the complete `messages + tools`, and (2) the sequence returned by slime's `qwen3` loss-mask generator and used for SFT. The sequences matched token-for-token for 123 trajectories and differed for 77. All 77 mismatches contained consecutive tool messages, while none of the other 123 did, exactly matching the trigger in the minimal reproduction.
